src/Apps/Order/Dialogs.tsx

```diff
@@ -0,0 +1,157 @@
+import { Link } from "@artsy/palette"
+import { ModalDialog, ModalDialogProps } from "Components/Modal/ModalDialog"
+import React from "react"
+import { Container, Subscribe } from "unstated"
+
+interface DialogState {
+  props: ModalDialogProps
+  forceCloseDialog: () => Promise<void>
+}
+
+export class DialogContainer extends Container<DialogState> {
+  state: DialogState = {
+    props: {
+      show: false,
+      heading: null,
+      detail: null,
+      primaryCta: {
+        text: "",
+        action: () => void 0,
+      },
+    },
+    forceCloseDialog: () => Promise.resolve(),
+  }
+
+  // closes the dialog without resolving consumers' promises
+  private closeDialog() {
+    return new Promise(resolve => {
+      this.setState({ props: { ...this.state.props, show: false } }, resolve)
+    })
+  }
+
+  // closes the dialog and resolves consumers' promises
+  // We are quite unlikely to need this, but let's err on the side of caution
+  private async maybeForceCloseExistingDialog() {
+    if (this.state.props.show) {
+      await this.state.forceCloseDialog()
+    }
+  }
+
+  showAcceptDialog = async ({
+    title,
+    message,
+    continueButtonText = "Continue",
+    cancelButtonText = "Cancel",
+  }: {
+    title: React.ReactNode
+    message: React.ReactNode
+    continueButtonText?: string
+    cancelButtonText?: string
+  }): Promise<{ accepted: boolean }> => {
+    await this.maybeForceCloseExistingDialog()
+
+    return new Promise<{ accepted: boolean }>(resolve => {
+      const accept = async () => {
+        await this.closeDialog()
+        resolve({ accepted: true })
+      }
+      const reject = async () => {
+        await this.closeDialog()
+        resolve({ accepted: false })
+      }
+
+      this.setState({
+        props: {
+          show: true,
+          heading: title,
+          detail: message,
+          primaryCta: {
+            text: continueButtonText,
+            action: accept,
+          },
+          secondaryCta: {
+            text: cancelButtonText,
+            action: reject,
+          },
+          onClose: reject,
+        },
```

Comment:
Nope. It doesn't even shallow diff.

Reply:
If not, I think this could lead to some confusion later. Might be better to just flatten things out or use something like [updeep](https://github.com/substantial/updeep) (though flattening would be better).

Reply:
Ha now I'm confused 🙃— what could be confusing later?

Reply:
Meaning, if someone later called (hypothetically) `this.setState({ props: { heading: "fool" }})` and no change occurred. But if it were flattened and `props` was removed then it would trigger an update as expected.

Reply:
I'm still not following. Calling `this.setState({ props: { heading: "foo" } })` would not work because `props` is of type `ModalDialogProps` and has a bunch of other required properties so TS would complain at you to add them. React doesn't merge state deeply, only at the top level.

Reply:
> React doesn't merge state deeply, only at the top level

That's what I'm saying -- namespacing the modal props under `props` breaks the default React state model since a change to a nested prop won't trigger an update (irrespective of TS validation). We should just remove the nested `props` object field and keep everything flat:

```js
state = {
  show: false,
  heading: null,
  detail: null,
  ... 
}
```

Reply:
> a change to a nested prop won't trigger an update

Seems like an update is always triggered when you call `setState` unless you try to replace the `state` object with `null` or `undefined` —  👉 https://codesandbox.io/s/62098oo6mk

Reply:
Ahh nice, thank you for this example 👍